### PR TITLE
Connector: getContentNodes for Github

### PR DESCRIPTION
**Pull request by PopDaph**
## Description

This PR implements the Content Nodes retriever for Github: `retrieveGithubReposContentNodes`.

On Github we can pick: 
- A full Repo
- All the Issues of a Repo
- All the Discussions of a Repo
- All the Code of a Repo
- A specific directly in the Code
- A specific file in the Code

I added a `titleWithParentsContext` to the type `ConnectorNode` to be able to display a specific name with more context, as displaying just "Code" or "Issues" can be very confusing if we don't display the name of the repo. 

Looks like this: 

Screenshot 2024-03-04 at 15 29 15

### CODE REVIEW

File: connectors/src/connectors/github/lib/utils.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  "REPO_CODE",
+  "REPO_CODE_DIR",
+  "REPO_CODE_FILE",
+] as const;
+export type GithubContentNodeType = (typeof GITHUB_CONTENT_NODE_TYPES)[number];
+
+/**
+ * Gets the type of the Github content node from its internal id.
+ */
+export function matchGithubInternalIdType(internalId: string): {
```
Comment (fontanierh) [inline connectors/src/connectors/github/lib/utils.ts:14]: Can you have an explicit return type here pls 🙏

File: connectors/src/connectors/github/index.ts [hunk 2]
```diff
@@ -459,6 +462,218 @@ export async function retrieveGithubReposTitles(
...
+      case "REPO_CODE":
+        allCodeFromRepoIds.push(repoId.toString());
+        break;
+      case "REPO_CODE_DIR":
+        codeDirectoryIds.push(internalId);
+        break;
+      case "REPO_CODE_FILE":
+        codeFileIds.push(internalId);
+        break;
+      default:
```
Comment (fontanierh) [inline connectors/src/connectors/github/index.ts:516]: Can you assertNever here pls 🙏  ?

File: connectors/src/connectors/github/index.ts [hunk 3]
```diff
@@ -459,6 +462,218 @@ export async function retrieveGithubReposTitles(
...
+    nodes.push({
+      provider: c.type,
+      internalId: repoId.toString(),
+      parentInternalId: null,
+      type: "folder",
+      title: repo.name,
+      titleWithParentsContext: `[${repo.name}] - Full repository`,
+      sourceUrl: repo.url,
+      expandable: true,
+      permission: "read" as ConnectorPermission,
```
Comment (fontanierh) [inline connectors/src/connectors/github/index.ts:570]: can you change this to not have `as ConnectorPermission` ? with the current code, if we ever change ConnectorPermission such that "read" is no longer valid, we won't get the type error here

File: connectors/src/connectors/github/index.ts [hunk 4]
```diff
@@ -459,6 +462,218 @@ export async function retrieveGithubReposTitles(
...
+    nodes.push({
+      provider: c.type,
+      internalId: `${repoId}-issues`,
+      parentInternalId: repoId.toString(),
+      type: "database",
+      title: "Issues",
+      titleWithParentsContext: `${repo.name} - Issues`,
+      sourceUrl: repo.url + "/issues",
+      expandable: false,
+      permission: "read" as ConnectorPermission,
```
Comment (fontanierh) [inline connectors/src/connectors/github/index.ts:591]: same here

**Review by fontanierh — COMMENTED** [review]
Looks great ! Just a few typescript-related comments

File: connectors/src/connectors/github/lib/utils.ts (hunk 1 above)
Comment (PopDaph) [reply connectors/src/connectors/github/lib/utils.ts:14]: Wait, I do already, or am I missing something? 😅

Comment (fontanierh) [reply connectors/src/connectors/github/lib/utils.ts:14]: indeed 🤦  My bad

File: connectors/src/connectors/github/index.ts (hunk 3 above)
Comment (PopDaph) [reply connectors/src/connectors/github/index.ts]: Done in all my function + on the code for getPermissions that was using the same!

Comment (fontanierh) [reply connectors/src/connectors/github/index.ts]: Awesome thank you !

**Review by fontanierh — APPROVED** [review]
thank you !

File: connectors/src/connectors/github/index.ts [hunk 5]
```diff
@@ -459,6 +462,218 @@ export async function retrieveGithubReposTitles(
...
+  codeDirectories.forEach((directory) => {
+    const repo = uniqueRepos[parseInt(directory.repoId)];
+    nodes.push({
+      provider: c.type,
+      internalId: directory.internalId,
+      parentInternalId: directory.parentInternalId,
+      type: "folder",
+      title: directory.dirName,
+      titleWithParentsContext: repo
+        ? `[${repo.name}] Code Directory "${directory.dirName}"`
```
Comment (spolu) [inline connectors/src/connectors/github/index.ts:644]: Sorry late to the party I would just remove "Code Directory" and "Code File" bellow. The type is covered by the icon and the fact that it comes from code almost obvious.

Maybe `[dust] README.md (code)` `[dust] srv (code)` if you care about keeping the code part.

Comment (PopDaph) [reply connectors/src/connectors/github/index.ts:644]: https://github.com/dust-tt/dust/pull/4119

**Pull request merged**